cli/src/commands/log.rs

```diff
@@ -76,27 +76,53 @@ pub(crate) fn cmd_log(
 ) -> Result<(), CommandError> {
     let workspace_command = command.workspace_helper(ui)?;
 
-    let revset_expression = {
-        let mut expression = if args.revisions.is_empty() {
+    ui.request_pager();
+    let revset = {
+        let expression = if args.revisions.is_empty() {
             workspace_command.parse_revset(&command.settings().default_revset())?
         } else {
             workspace_command.parse_union_revsets(&args.revisions)?
         };
-        if !args.paths.is_empty() {
+        if args.paths.is_empty() {
+            workspace_command.evaluate_revset(expression)?
+        } else {
             let repo_paths: Vec<_> = args
                 .paths
                 .iter()
                 .map(|path_arg| workspace_command.parse_file_path(path_arg))
                 .try_collect()?;
-            expression = expression.intersection(&RevsetExpression::filter(
-                RevsetFilterPredicate::File(Some(repo_paths)),
-            ));
+            let path_filter =
+                &RevsetExpression::filter(RevsetFilterPredicate::File(Some(repo_paths)));
+
+            let revset = workspace_command.evaluate_revset(expression.intersection(path_filter))?;
+            if !revset.is_empty() {
+                revset
+            } else {
+                let revset_all = workspace_command
+                    .evaluate_revset(RevsetExpression::all().intersection(path_filter))?;
+                if revset_all.is_empty() {
+                    writeln!(
+                        ui.warning_default(),
+                        "The repository contains no changes for the provided paths."
+                    )?;
+                    revset
+                } else {
+                    writeln!(
+                        ui.warning_default(),
+                        "The revset '{}' contains no changes for the provided paths, using 'all()' instead.",
+                        if args.revisions.is_empty() {
+                            command.settings().default_revset()
+                        } else {
+                            args.revisions.iter().map(|rev| rev.to_string()).join(" | ")
+                        }
```

Comment:
Not sure if this part is even necessary at all, but I think it might be helpful to print out the revset that was used when no changes were found. If this is applicable to other commands, it might make sense to implement Display for RevsetExpression which would make it very easy to print out the used revset whether it was created from user input or programatically.